### ISSUE TITLE
Allow throw statement for eslint no-unused-expressions with allowShortCircuit=true

### ISSUE THREAD

**Issue by PatriceVignola** [eslint/eslint#10382]
I don't know if this has been discussed before or if it's intended behavior, but I was surprised when `no-unused-expressions` with `allowShortCircuit=true` was flagging the following lines as an error:

```js
/*eslint no-unused-expressions: ["error", { "allowShortCircuit": true }]*/
a && throw Error('Short Error');
a || throw Error('Short error');
```

It might be due to the fact that `throw` is a keyword instead of a function, but it's still technically a short circuit. It's very useful and clean to be able to do this instead of having an if statement every time.

I can make a PR myself if it's a wanted change.

**What rule do you want to change?**
no-unused-expressions

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Change the behavior of the allowShortCircuit option to allow throw statements.

**Please provide some example code that this change will affect:**

```js
/*eslint no-unused-expressions: ["error", { "allowShortCircuit": true }]*/

// Valid code
a && throw Error('Short Error');
```

**What does the rule currently do for this code?**
It throws a `no-unused-expressions` error, even if `allowShortCircuit=true`.

**What will the rule do after it's changed?**
It will not throw a `no-unused-expressions` error anymore when `allowShortCircuit=true`.

**Comment by PatriceVignola** [comment]
Ah, my bad, I should have tested it properly. Flow wasn't throwing an error and my code was never actually reaching that line so I thought it was valid syntax. I restarted the Flow server and it properly flags it to me now.  Thanks! I'll close the issue.

**Comment by platinumazure** [comment]
Thanks! Reason for the parse error is that (at least in non-experimental JS) throw statements are statements, not expressions, so it's impossible to embed them in an expression.
